rs/ic-observability/multiservice-discovery/src/definition.rs

```diff
@@ -669,3 +670,72 @@ impl DefinitionsSupervisor {
         Ok(())
     }
 }
+
+#[derive(Deserialize)]
+pub struct TargetFilterSpec {
+    pub node_provider_id: Option<String>,
+    pub operator_id: Option<String>,
+    pub dc_id: Option<String>,
+    pub ic_name: Option<String>,
+    pub subnet_id: Option<String>,
+}
+
+impl TargetFilterSpec {
+    pub fn matches_ic_node(&self, t: &TargetDto) -> bool {
+        // self.ic_name is explicitly excluded here.
+        // Call self.matches_ic().
+        let o = match &self.operator_id {
+            None => true,
+            Some(operator_id) => t.operator_id.to_string() == operator_id.to_owned(),
+        };
+        let n = match &self.node_provider_id {
+            None => true,
+            Some(node_provider_id) => t.node_provider_id.to_string() == node_provider_id.to_owned(),
+        };
+        let d = match &self.dc_id {
+            None => true,
+            Some(dc_id) => t.dc_id.to_string() == dc_id.to_owned(),
+        };
+        let s = match &self.subnet_id {
+            None => true,
+            Some(subnet_id) => match t.subnet_id {
+                Some(t_subnet_id) => t_subnet_id.to_string() == subnet_id.to_owned(),
+                None => subnet_id.as_str() == "",
+            },
+        };
+        o && n && d && s
+    }
+
+    pub fn matches_boundary_node(&self, b: &BoundaryNode) -> bool {
+        // self.ic_name is explicitly excluded here.
+        // Call self.matches_ic().
+        if self.operator_id.is_some() || self.node_provider_id.is_some() || self.subnet_id.is_some() {
+            return true;
+        };
```

Comment:
Is this what we want? If any of these fields are provided, the BN filtering will "match" without and further checks.
What if NP + DC filter is provided? The function will return all BNs.

Reply:
Whoa, what a wonderful catch.  It turns out it's meant to be return false.